Added `has_children` variable to category tags, #1726

robsonsobral opened the pull request:
<!-- What's in this pull request? -->
## Overview

Add `{has_children}` variable to `exp:channel:categories` and `exp:channel:category_archive`.

<!-- If this pull request resolves a project issue, provide a link: -->
Resolves [#1726](https://github.com/ExpressionEngine/ExpressionEngine/issues/1726).

## Nature of This Change

<!-- Check all that apply: -->

- [ ] 🐛 Fixes a bug
- [x] 🚀 Implements a new feature
- [ ] 🛁 Refactors existing code
- [ ] 💅 Fixes coding style
- [ ] ✅ Adds tests
- [ ] 👽 Adds new dependency
- [ ] 🔥 Removes unused files / code
- [ ] 🔒 Improves security <!-- if your fix would EXPOSE a current security flaw, do not submit a pull request. Instead report a security bug at https://docs.expressionengine.com/latest/bugs_and_security_reports -->

## Is this backwards compatible?

- [x] Yes
- [ ] No

## Documentation
<!-- Required for new features -->
User Guide Pull Request: https://github.com/ExpressionEngine/ExpressionEngine-User-Guide/pull/379

robsonsobral committed:
Add `has_children` to `exp:categories` and `exp:category_archive`

intoeetive reviewed (dismissed):
The code looks good to me, but need documentation (and preferably tests)

robsonsobral commented:
Tests added in the dark, due to environment issues.

I started to test for `{if active}` too, but I got lost.

robsonsobral commented:
@intoeetive , I wasn't capable of setting the parent category of **category two** without seeing what I'm doing.

intoeetive commented:
It does not looks like tests have been written correctly. You're expecting `category one` to have children, however it does not have any

robsonsobral commented:
Thank you, @intoeetive . Everything is working, now.

I updated some previous tests to also consider category nesting.

intoeetive commented on tests/cypress/support/templates/cats.group/index.html:
```diff
@@ -1,4 +1,4 @@
-{exp:channel:categories channel="news" style="linear"}
```
Linear and nested categories use different code that generates display. This test template is specifically set to use linear mode, so please let's continue using it. And maybe we need separate test template for the nested mode.

intoeetive reviewed (changes requested):
Spotted an issue in test templates.

If you still can't get Cypress running locally, I can try making that update for you

robsonsobral commented:
For reviewers to keep an eye on:

- the conflicts are related to a new test being added. Could it have a better description?

intoeetive reviewed (commented):
@robsonsobral the code looks good to me!

But I have question on tests - it looks like you're changing the "reorder" test to nest categories instead of reordering. Would it be better to add new test instead? Basically, keep the old and new version

intoeetive commented:
@robsonsobral I'm making this PR draft until conflict on tests is resolved (I still think we just need a separate test for this change, and keep the existing test for other things)

robsonsobral commented:
I've already move back your tests to the original state, @intoeetive , and just added mine. However, it wasn't clear among all changes. So, I rebased this branch on the current dev. Everything is clearer now.

I gonna move my tests to their own `it()`s tomorrow (for me).

robsonsobral commented:
@intoeetive , the templates for tests aren't being copied to the system folder. Weird...

robsonsobral committed:
Add tests

robsonsobral commented:
I've no idea why the build tools are failing!

robsonsobral commented:
Is there something I can do to fix the **Build EE for testing**? There's an error on `gulp`:

> Run gulp build-app --local --nogit --head --skip-lint --skip-pro --version=6.3.4
> ReferenceError: primordials is not defined

intoeetive committed:
modify the templates for this test
